packages/headless/src/features/folding/folding-slice.ts

```diff
@@ -33,70 +39,103 @@ function areDefinedAndEqual<T>(
   return (value1 || value2) !== undefined && value1 === value2;
 }
 
-function flattenResultTree(result: ResultWithFolding): ResultWithFolding[] {
-  const results = [result];
-  if (result.parentResult) {
-    results.unshift(result.parentResult);
-  }
-  if (result.childResults?.length) {
-    results.push(
-      ...result.childResults.flatMap((childResult) =>
-        flattenResultTree(childResult)
-      )
-    );
+function getChildResultsRecursively(
+  result: ResultWithFolding
+): ResultWithFolding[] {
+  if (!result.childResults) {
+    return [];
   }
-  return results;
+  return result.childResults.flatMap((childResult) => [
+    childResult,
+    ...getChildResultsRecursively(childResult),
+  ]);
 }
 
-function foldResult(
-  source: ResultWithFolding,
+function resolveChildrenFromFields(
+  parent: ResultWithFolding,
   results: ResultWithFolding[],
   fields: FoldingFields
-): FoldedResult {
-  const sourceChildValue = getChildField(source, fields);
-  return {
-    ...source,
-    children: sourceChildValue
-      ? results
-          .filter((result) => {
-            const isSameResultAsSource =
-              getChildField(result, fields) === getChildField(source, fields);
-            const isChildOfSource =
-              getParentField(result, fields) === sourceChildValue;
-            return isChildOfSource && !isSameResultAsSource;
-          })
-          .map((result) => foldResult(result, results, fields))
-      : [],
-  };
+): FoldedResult[] {
+  const sourceChildValue = getChildField(parent, fields);
+  return sourceChildValue
+    ? results
+        .filter((result) => {
+          const isSameResultAsSource =
+            getChildField(result, fields) === getChildField(parent, fields);
+          const isChildOfSource =
+            getParentField(result, fields) === sourceChildValue;
+          return isChildOfSource && !isSameResultAsSource;
+        })
+        .map((result) => ({
+          result,
+          children: resolveChildrenFromFields(result, results, fields),
+        }))
+    : [];
 }
 
-function createCollection(
-  relevantResult: ResultWithFolding,
+function resolveRootFromFields(
```

Comment:
Could we use the terminology parent/children instead of root/children?

Reply:
Or a children of the root could also be a parent... hmm that case forget what I said, I'd use RootParent perhaps

Reply:
I think `root` is more accurate here though, since we're resolving the top-most parent and I feel like `parent` is a relative term while `root` is absolute.